### PR TITLE
Seperate segmented algorithms for adjacent_find

### PR DESCRIPTION
**Pull request by Jedi18**
Seperate segmented algorithms for adjacent_find

Any background context you want to provide?
Issue #5156

### CODE REVIEW

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/adjacent_find.hpp [hunk 1]
```diff
@@ -348,12 +347,16 @@ namespace hpx { namespace ranges {
...
                 typename hpx::traits::range_traits<Rng>::iterator_type;
-            using is_segmented =
-                hpx::traits::is_segmented_iterator<iterator_type>;
-            return hpx::parallel::v1::detail::adjacent_find_(
-                hpx::execution::seq, std::begin(rng), std::end(rng),
-                std::forward<Pred>(pred), std::forward<Proj>(proj),
-                is_segmented());
+
+            static_assert(
+                (hpx::traits::is_forward_iterator<iterator_type>::value),
```
Comment (Jedi18) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/adjacent_find.hpp:352]: Should I change this to input iterator?

Comment (aurianer) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/adjacent_find.hpp:352]: Nope if you check the [C++ working draft](http://eel.is/c++draft/alg.adjacent.find), it is a forward iterator for adjacent_find

Comment (Jedi18) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/adjacent_find.hpp:352]: Ohh ok thanks! I'll be sure to refer the C++ working draft from next time.

**Comment by aurianer** [comment]
retest

**Comment by hkaiser** [comment]
retest cscs

**Review by aurianer — DISMISSED** [review]
LGTM thanks a lot! 2 minor suggestions but it's not a big deal if it doesn't go into this PR :)

File: libs/full/segmented_algorithms/include/hpx/parallel/segmented_algorithms/adjacent_find.hpp [hunk 2]
```diff
@@ -274,39 +274,77 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+        hpx::adjacent_find_t, InIter first, InIter last, Pred&& pred = Pred())
+    {
+        static_assert((hpx::traits::is_input_iterator<InIter>::value),
+            "Requires at least input iterator.");
+
+        if (first == last)
         {
-            using is_seq = hpx::is_sequenced_execution_policy<ExPolicy>;
-            using result = util::detail::algorithm_result<ExPolicy, FwdIter>;
+            return last;
```
[the remark this answers is not in the document]
Comment (hkaiser) [reply libs/full/segmented_algorithms/include/hpx/parallel/segmented_algorithms/adjacent_find.hpp:300]: Excellent catch @aurianer. This is definitely something that needs fixing for this PR.

[the remark this answers is not in the document]
Comment (Jedi18) [reply libs/full/segmented_algorithms/include/hpx/parallel/segmented_algorithms/adjacent_find.hpp]: Thanks! I've made the changes.

**Review by hkaiser — APPROVED** [review]
Thanks!

**Review by aurianer — APPROVED** [review]
LGTM thanks!

**Pull request merged**